### PR TITLE
Fixes #25053 - expensive db_facts query executed just once

### PR DESCRIPTION
**Pull request by lzap**
It looks like this query is very expensive on MySQL: `host.fact_values.where(:fact_name => fact_names.values)`

We call it twice currently while it is not needed to run this query twice, we can fetch the list of fact value IDs stored in database in update step and prepare list of facts to be added in the next step in advance.

Also we order the query by ID for some reason, we can process them unordered just fine.

### CODE REVIEW

**Comment by lzap** [comment]
The current version does not help, we actually need to get rid of `find_each`, we can replace with `pluck` in one case and the other case does not matter because we load all records into memory.

**Comment by lzap** [comment]
Here is new version, I simply replaced `find_each` with just `each`. It turned out that `find_each` creates queries with `ORDER BY id LIMIT 1000` for batch processing which is causing problems in production. Although `find_each` is more memory effective, this was biting us only for entries with huge fact lists which we now prevent using fact filters.

**Comment by mmoll** [comment]
test failures (partly) related?

**Comment by tbrisker** [comment]
I'm a bit concerned about the memory impact of this. have you checked the memory affect of this change with very large facts tables?

**Comment by lzap** [comment]
For the record, there is a more simple alternative of the patch: https://github.com/theforeman/foreman/pull/6108

**Comment by lzap** [comment]
@tbrisker I agree for the first case, the `initialize_fact_names` method should be fine tho - the purpose of this method is to actually load all records into memory as part of optimalization we've done.

For the latter occurrence, I am falling back to `find_each` but providing USE INDEX statement for MySQL connections. Let's see what Jenkins says.

**Comment by LangJV** [comment]
It looks like the checks now pass for mysql and sqllite3 - but fail for postgresql on org and location specific checks, if im interpreting the jenkins matrix correctly. Looking at the code, i'm not seeing anything that would alter the returns in terms of organization and location (they arent referenced at all tbh from what im seeing..)

**Comment by mmoll** [comment]
[test foreman]

**Comment by mmoll** [comment]
:green_heart:

**Comment by lzap** [comment]
Wow it's a pass. Cool! Edit: To elaborate for @LangJV we sometimes get random failures on Jenkins.

**Comment by LangJV** [comment]
I've gotten this through both my nonprod and prod environments. Seems to be working perfectly, no other errors I'm noticing. The db load has gone from 34 (16 core box) down to 2-5. I'm hopeful i can shrink this box back to 8 cores again now :slight_smile:

**Review by mmoll — APPROVED** [review]
I think this is good and should go also to 1.19 and 1.18

File: app/services/fact_importer.rb [hunk 1]
```diff
@@ -161,23 +161,25 @@ def add_new_fact(name)
   end
 
   def add_new_facts
-    facts_to_create = facts.keys - db_facts.pluck(:fact_name_id).map { |name_id| fact_names_by_id[name_id].name }
```
Comment (tbrisker) [inline app/services/fact_importer.rb:164]: Doesn't this mean we no longer need to store the `fact_names_by_id` hash?

**Review by tbrisker — APPROVED** [review]
This is already a good improvement, I'll merge this now and remove the extra hash in another PR.

**Comment by tbrisker** [comment]
https://github.com/theforeman/foreman/pull/6130

**Comment by tbrisker** [comment]
@mmoll I'll pull this back to 1.19 with the other one once it is merged.

File: app/services/fact_importer.rb (hunk 1 above)
Comment (lzap) [reply app/services/fact_importer.rb:164]: I wanted to be absolutely compatible with plugins which build on top of this class. But neither Ansible or Salt uses this, we shall consider removing it. Probably a Ref PR that would NOT go into 1.20 (so we merge it later on after branching)?

Comment (tbrisker) [reply app/services/fact_importer.rb:164]: These methods are all private, so they shouldn't be used in any plugins, which is why I'm fine with completely removing the hash with no deprecation.

**Pull request merged**